### PR TITLE
Implement CAMERA_VISIBLE_LAYERS as built-in shader variable

### PR DESCRIPTION
**Pull request by NumbuhFour**
Adds a built-in variable for shaders which exposes the cull mask of the camera rendering the current pass.

Godot_VisibleLayers_Shader_example_pr
In this example, each camera has a different cull mask and the cube's shader outputs a different albedo for each.

Implements godotengine/godot-proposals#5563
Tested with render methods forward_plus, gl_compatibility, and mobile

### CODE REVIEW

**Review by mhilbrunner — COMMENTED** [review]
PR meeting: discussed this, looks good, the feature is wanted, needs review from @reduz and @clayjohn

File: drivers/gles3/shaders/scene.glsl [hunk 1]
```diff
@@ -129,7 +129,7 @@ layout(std140) uniform SceneData { // ubo:2
 
 	mediump float ambient_color_sky_mix;
 	bool material_uv2_mode;
-	float pad2;
+	float emissive_exposure_normalization;
```
Comment (clayjohn) [inline drivers/gles3/shaders/scene.glsl:132]: Good catch here, these should have been renamed when Physical Light Units were added

File: servers/rendering/renderer_scene_render.h [hunk 2]
```diff
@@ -278,7 +279,7 @@ class RendererSceneRender {
 		Projection view_projection[RendererSceneRender::MAX_RENDER_VIEWS];
 		Vector2 taa_jitter;
 
-		void set_camera(const Transform3D p_transform, const Projection p_projection, bool p_is_orthogonal, bool p_vaspect, const Vector2 &p_taa_jitter = Vector2());
+		void set_camera(const Transform3D p_transform, const Projection p_projection, bool p_is_orthogonal, bool p_vaspect, const Vector2 &p_taa_jitter = Vector2(), uint32_t p_visible_layers = 0x00000000);
```
Comment (clayjohn) [inline servers/rendering/renderer_scene_render.h:282]:
```suggestion
		void set_camera(const Transform3D p_transform, const Projection p_projection, bool p_is_orthogonal, bool p_vaspect, const Vector2 &p_taa_jitter = Vector2(), uint32_t p_visible_layers = 0xFFFFFFFF);
```
Default should be ``0xFFFFFFFF`` to match the default in renderer_scene_cull.h

File: servers/rendering/shader_types.cpp [hunk 3]
```diff
@@ -100,6 +100,7 @@ ShaderTypes::ShaderTypes() {
 	shader_modes[RS::SHADER_SPATIAL].functions["vertex"].built_ins["NODE_POSITION_WORLD"] = ShaderLanguage::TYPE_VEC3;
 	shader_modes[RS::SHADER_SPATIAL].functions["vertex"].built_ins["CAMERA_POSITION_WORLD"] = ShaderLanguage::TYPE_VEC3;
 	shader_modes[RS::SHADER_SPATIAL].functions["vertex"].built_ins["CAMERA_DIRECTION_WORLD"] = ShaderLanguage::TYPE_VEC3;
+	shader_modes[RS::SHADER_SPATIAL].functions["vertex"].built_ins["CAMERA_VISIBLE_LAYERS"] = ShaderLanguage::TYPE_INT;
```
Comment (clayjohn) [inline servers/rendering/shader_types.cpp:103]: Should be UINT
```suggestion
	shader_modes[RS::SHADER_SPATIAL].functions["vertex"].built_ins["CAMERA_VISIBLE_LAYERS"] = ShaderLanguage::TYPE_UINT;
```

File: servers/rendering/shader_types.cpp [hunk 4]
```diff
@@ -147,6 +148,7 @@ ShaderTypes::ShaderTypes() {
 	shader_modes[RS::SHADER_SPATIAL].functions["fragment"].built_ins["NODE_POSITION_WORLD"] = ShaderLanguage::TYPE_VEC3;
 	shader_modes[RS::SHADER_SPATIAL].functions["fragment"].built_ins["CAMERA_POSITION_WORLD"] = ShaderLanguage::TYPE_VEC3;
 	shader_modes[RS::SHADER_SPATIAL].functions["fragment"].built_ins["CAMERA_DIRECTION_WORLD"] = ShaderLanguage::TYPE_VEC3;
+	shader_modes[RS::SHADER_SPATIAL].functions["fragment"].built_ins["CAMERA_VISIBLE_LAYERS"] = ShaderLanguage::TYPE_INT;
```
Comment (clayjohn) [inline servers/rendering/shader_types.cpp:151]:
```suggestion
	shader_modes[RS::SHADER_SPATIAL].functions["fragment"].built_ins["CAMERA_VISIBLE_LAYERS"] = ShaderLanguage::TYPE_UINT;
```

File: servers/rendering/renderer_scene_render.cpp [hunk 5]
```diff
@@ -49,6 +50,7 @@ void RendererSceneRender::CameraData::set_camera(const Transform3D p_transform,
 void RendererSceneRender::CameraData::set_multiview_camera(uint32_t p_view_count, const Transform3D *p_transforms, const Projection *p_projections, bool p_is_orthogonal, bool p_vaspect) {
 	ERR_FAIL_COND_MSG(p_view_count != 2, "Incorrect view count for stereoscopic view");
 
+	visible_layers = 0x00000000;
```
Comment (clayjohn) [inline servers/rendering/renderer_scene_render.cpp:53]:
```suggestion
	visible_layers = 0xFFFFFFFF;
```

File: servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h [hunk 6]
```diff
@@ -142,7 +143,7 @@ class RenderSceneDataRD {
 		float IBL_exposure_normalization; // Adjusts for baked exposure.
 
 		uint32_t pancake_shadows;
-		uint32_t pad1;
+		int32_t camera_visible_layers;
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h:146]:
```suggestion
		uint32_t camera_visible_layers;
```

File: servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h [hunk 7]
```diff
@@ -45,6 +45,7 @@ class RenderSceneDataRD {
 	Transform3D cam_transform;
 	Projection cam_projection;
 	Vector2 taa_jitter;
+	int32_t camera_visible_layers;
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.h:48]:
```suggestion
	uint32_t camera_visible_layers;
```

File: servers/rendering/renderer_rd/shaders/scene_data_inc.glsl [hunk 8]
```diff
@@ -63,7 +63,7 @@ struct SceneData {
 	float IBL_exposure_normalization;
 
 	bool pancake_shadows;
-	uint pad1;
+	int camera_visible_layers;
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/scene_data_inc.glsl:66]:
```suggestion
	uint camera_visible_layers;
```

File: servers/rendering/renderer_rd/renderer_scene_render_rd.cpp [hunk 9]
```diff
@@ -1078,6 +1078,8 @@ void RendererSceneRenderRD::render_scene(const Ref<RenderSceneBuffers> &p_render
 		scene_data.cam_transform = p_camera_data->main_transform;
 		scene_data.cam_projection = p_camera_data->main_projection;
 		scene_data.cam_orthogonal = p_camera_data->is_orthogonal;
+		// There's no shaderport for uints, and only 16 layers anyway. So drop sign bit
+		scene_data.camera_visible_layers = (int)p_camera_data->visible_layers;
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:1082]:
```suggestion
		scene_data.camera_visible_layers = p_camera_data->visible_layers;
```
Its simpler to just consistently use uints

File: servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp [hunk 10]
```diff
@@ -701,6 +701,8 @@ void SceneShaderForwardClustered::init(const String p_defines) {
 		actions.renames["CAMERA_POSITION_WORLD"] = "scene_data.inv_view_matrix[3].xyz";
 		actions.renames["CAMERA_DIRECTION_WORLD"] = "scene_data.view_matrix[3].xyz";
 		actions.renames["NODE_POSITION_VIEW"] = "(read_model_matrix * scene_data.view_matrix)[3].xyz";
+		actions.renames["CAMERA_VISIBLE_LAYERS"] = "scene_data.camera_visible_layers";
+		actions.renames["NODE_POSITION_VIEW"] = "(model_matrix * scene_data.view_matrix)[3].xyz";
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp:705]:
```suggestion
		actions.renames["CAMERA_VISIBLE_LAYERS"] = "scene_data.camera_visible_layers";
```
This looks like a mistake from rebasing ``NODE_POSITION_VIEW`` is already defined above

File: servers/rendering/renderer_rd/forward_mobile/scene_shader_forward_mobile.cpp [hunk 11]
```diff
@@ -601,6 +601,8 @@ void SceneShaderForwardMobile::init(const String p_defines) {
 		actions.renames["CAMERA_POSITION_WORLD"] = "scene_data.inv_view_matrix[3].xyz";
 		actions.renames["CAMERA_DIRECTION_WORLD"] = "scene_data.view_matrix[3].xyz";
 		actions.renames["NODE_POSITION_VIEW"] = "(read_model_matrix * scene_data.view_matrix)[3].xyz";
+		actions.renames["CAMERA_VISIBLE_LAYERS"] = "scene_data.camera_visible_layers";
+		actions.renames["NODE_POSITION_VIEW"] = "(model_matrix * scene_data.view_matrix)[3].xyz";
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/forward_mobile/scene_shader_forward_mobile.cpp:605]:
```suggestion
		actions.renames["CAMERA_VISIBLE_LAYERS"] = "scene_data.camera_visible_layers";
```

File: drivers/gles3/shaders/scene.glsl [hunk 12]
```diff
@@ -151,6 +151,10 @@ layout(std140) uniform SceneData { // ubo:2
 
 	vec3 fog_light_color;
 	float fog_sun_scatter;
+	int camera_visible_layers;
```
Comment (clayjohn) [inline drivers/gles3/shaders/scene.glsl:154]:
```suggestion
	uint camera_visible_layers;
```

File: drivers/gles3/shaders/scene.glsl [hunk 13]
```diff
@@ -475,6 +479,10 @@ layout(std140) uniform SceneData { // ubo:2
...
```
Comment (clayjohn) [inline drivers/gles3/shaders/scene.glsl:482]:
```suggestion
	uint camera_visible_layers;
```

File: drivers/gles3/rasterizer_scene_gles3.h [hunk 14]
```diff
@@ -103,6 +103,7 @@ struct RenderDataGLES3 {
 	Transform3D inv_cam_transform = Transform3D();
 	Projection cam_projection = Projection();
 	bool cam_orthogonal = false;
+	int32_t camera_visible_layers = 0x00000000;
```
Comment (clayjohn) [inline drivers/gles3/rasterizer_scene_gles3.h:106]:
```suggestion
	uint32_t camera_visible_layers = 0xFFFFFFFF;
```
Should be a uint32_t and the initialization should match the one in renderer_scene_cull.h

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 15]
```diff
@@ -1653,6 +1654,8 @@ void RasterizerSceneGLES3::render_scene(const Ref<RenderSceneBuffers> &p_render_
 		render_data.inv_cam_transform = render_data.cam_transform.affine_inverse();
 		render_data.cam_projection = p_camera_data->main_projection;
 		render_data.cam_orthogonal = p_camera_data->is_orthogonal;
+		// There's no shaderport for uints, and only 16 layers anyway. So drop sign bit
+		render_data.camera_visible_layers = (int)p_camera_data->visible_layers;
```
Comment (clayjohn) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1658]:
```suggestion
		render_data.camera_visible_layers = p_camera_data->visible_layers;
```

**Review by clayjohn — COMMENTED** [review]
Looks mostly fine! I have left comments on places where uints should be used in place of ints. In my mind it is preferable to be consistent and use ``uint32_t``s throughout instead of switching to ``int``s when passing the data to the renderer. ``uints`` are supported in both GLSL 330 and 450 so using them should be fine in all shaders.

Also the default should be the same as the default in renderer_scene_cull.h

**Comment by NumbuhFour** [comment]
> uints should be used in place of ints. In my mind it is preferable to be consistent and use `uint32_t`s throughout instead of switching to `int`s when passing the data to the renderer. `uints` are supported in both GLSL 330 and 450 so using them should be fine in all shaders.

@clayjohn I originally made that decision because I thought godot's shader translation couldn't handle uints as I was getting errors, but I have discovered that it was just that uints have their own literal suffix, so all your suggested changes have been implemented. Thank you for the review!

An example shader now:
```
uint layers = CAMERA_VISIBLE_LAYERS;
if ((layers & 3u) == 3u) { // Both layers 1 and 2, render blue
	ALBEDO = vec3(0.0,0.0,1.0);
}
else if ((layers & 1u) == 1u) { // Only layer 1, render red
	ALBEDO = vec3(1.0,0.0,0.0);
}
else if ((layers & 2u) == 2u) { // Only layer 2, render green
	ALBEDO = vec3(0.0,1.0,0.0);
}
```

File: drivers/gles3/rasterizer_scene_gles3.h [hunk 16]
```diff
@@ -343,6 +344,10 @@ class RasterizerSceneGLES3 : public RendererSceneRender {
 
 			float fog_light_color[3];
 			float fog_sun_scatter;
+			int32_t camera_visible_layers;
```
[the remark this answers is not in the document]
Comment (NumbuhFour) [reply drivers/gles3/rasterizer_scene_gles3.h]: Thank you for being so thorough, sorry I missed these.

**Review by clayjohn — APPROVED** [review]

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
This looks ok to me.

File: scene/resources/visual_shader.cpp [hunk 17]
```diff
@@ -2658,6 +2658,7 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::ports[] = {
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_VERTEX, VisualShaderNode::PORT_TYPE_VECTOR_3D, "node_position_world", "NODE_POSITION_WORLD" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_VERTEX, VisualShaderNode::PORT_TYPE_VECTOR_3D, "camera_position_world", "CAMERA_POSITION_WORLD" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_VERTEX, VisualShaderNode::PORT_TYPE_VECTOR_3D, "camera_direction_world", "CAMERA_DIRECTION_WORLD" },
+	{ Shader::MODE_SPATIAL, VisualShader::TYPE_VERTEX, VisualShaderNode::PORT_TYPE_SCALAR_INT, "camera_visible_layers", "CAMERA_VISIBLE_LAYERS" },
```
Comment (Chaosus) [inline scene/resources/visual_shader.cpp:2661]:
```suggestion
	{ Shader::MODE_SPATIAL, VisualShader::TYPE_VERTEX, VisualShaderNode::PORT_TYPE_SCALAR_INT, "camera_visible_layers", "int(CAMERA_VISIBLE_LAYERS)" },
```

File: scene/resources/visual_shader.cpp [hunk 18]
```diff
@@ -2690,6 +2691,7 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::ports[] = {
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_3D, "node_position_world", "NODE_POSITION_WORLD" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_3D, "camera_position_world", "CAMERA_POSITION_WORLD" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_3D, "camera_direction_world", "CAMERA_DIRECTION_WORLD" },
+	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SCALAR_INT, "camera_visible_layers", "CAMERA_VISIBLE_LAYERS" },
```
Comment (Chaosus) [inline scene/resources/visual_shader.cpp:2694]:
```suggestion
	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SCALAR_INT, "camera_visible_layers", "int(CAMERA_VISIBLE_LAYERS)" },
```

Same

File: scene/resources/visual_shader.cpp (hunk 17 above)
Comment (NumbuhFour) [reply scene/resources/visual_shader.cpp:2661]: Good catch. Integrated this and confirmed working in visual shaders

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**